File: win-prog/ProgWinCS/Lines Curves and Area Fills/RoundRect/RoundRect.cs
//----------------------------------------
// RoundRect.cs © 2001 by Charles Petzold
//----------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class RoundRect: PrintableForm
{
     public new static void Main()
     {
          Application.Run(new RoundRect());
     }
     public RoundRect()
     {
          Text = "Rounded Rectangle";
     }
     protected override void DoPage(Graphics grfx, Color clr, int cx, int cy)
     {
          RoundedRectangle(grfx, Pens.Red, 
                           new Rectangle(0, 0, cx - 1, cy - 1),
                           new Size(cx / 5, cy / 5));
     }
     void RoundedRectangle(Graphics grfx, Pen pen, Rectangle rect, Size size)
     {
          grfx.DrawLine(pen, rect.Left  + size.Width / 2, rect.Top,
                             rect.Right - size.Width / 2, rect.Top);

          grfx.DrawArc(pen, rect.Right - size.Width, rect.Top, 
                            size.Width, size.Height, 270, 90);

          grfx.DrawLine(pen, rect.Right, rect.Top + size.Height / 2,
                             rect.Right, rect.Bottom - size.Height / 2);

          grfx.DrawArc(pen, rect.Right - size.Width, 
                            rect.Bottom - size.Height,
                            size.Width, size.Height, 0, 90);

          grfx.DrawLine(pen, rect.Right - size.Width / 2, rect.Bottom,
                             rect.Left + size.Width / 2, rect.Bottom);

          grfx.DrawArc(pen, rect.Left, rect.Bottom - size.Height,
                            size.Width, size.Height, 90, 90);

          grfx.DrawLine(pen, rect.Left, rect.Bottom - size.Height / 2,
                             rect.Left, rect.Top + size.Height / 2);

          grfx.DrawArc(pen, rect.Left, rect.Top,
                            size.Width, size.Height, 180, 90);
     }
}

File: win-prog/ProgWinVB/Lines Curves and Area Fills/RoundRect/RoundRect.vb
'------------------------------------------
' RoundRect.vb (c) 2002 by Charles Petzold
'------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class RoundRect
    Inherits PrintableForm

    Shared Shadows Sub Main()
        Application.Run(New RoundRect())
    End Sub

    Sub New()
        Text = "Rounded Rectangle"
    End Sub

    Protected Overrides Sub DoPage(ByVal grfx As Graphics, _
            ByVal clr As Color, ByVal cx As Integer, ByVal cy As Integer)
        RoundedRectangle(grfx, Pens.Red, _
                         New Rectangle(0, 0, cx - 1, cy - 1), _
                         New Size(cx \ 5, cy \ 5))
    End Sub

    Private Sub RoundedRectangle(ByVal grfx As Graphics, ByVal pn As Pen, _
                     ByVal rect As Rectangle, ByVal sz As Size)
        grfx.DrawLine(pn, rect.Left + sz.Width \ 2, rect.Top, _
                          rect.Right - sz.Width \ 2, rect.Top)
        grfx.DrawArc(pn, rect.Right - sz.Width, rect.Top, _
                         sz.Width, sz.Height, 270, 90)
        grfx.DrawLine(pn, rect.Right, rect.Top + sz.Height \ 2, _
                          rect.Right, rect.Bottom - sz.Height \ 2)
        grfx.DrawArc(pn, rect.Right - sz.Width, _
                         rect.Bottom - sz.Height, _
                         sz.Width, sz.Height, 0, 90)
        grfx.DrawLine(pn, rect.Right - sz.Width \ 2, rect.Bottom, _
                          rect.Left + sz.Width \ 2, rect.Bottom)
        grfx.DrawArc(pn, rect.Left, rect.Bottom - sz.Height, _
                         sz.Width, sz.Height, 90, 90)
        grfx.DrawLine(pn, rect.Left, rect.Bottom - sz.Height \ 2, _
                          rect.Left, rect.Top + sz.Height \ 2)
        grfx.DrawArc(pn, rect.Left, rect.Top, _
                         sz.Width, sz.Height, 180, 90)
    End Sub
End Class
